configs/components/openssl-fips.rb

```diff
@@ -0,0 +1,56 @@
+component "openssl-fips" do |pkg, settings, platform|
+
+  pkg.build_requires "puppet-runtime"
+
+  openssl_fips_details = JSON.parse(File.read('configs/components/openssl-fips.json'))
+  openssl_fips_version = openssl_fips_details['version']
+  openssl_fips_location = openssl_fips_details['location']
+
+  tarball_name = "#{pkg.get_name}-#{openssl_fips_version}.#{platform.name}.tar.gz"
+  pkg.url File.join(openssl_fips_location, tarball_name)
+  pkg.sha1sum File.join(openssl_fips_location, "#{tarball_name}.sha1")
+  pkg.version openssl_fips_version
+  pkg.install_only true
+
+  pkg.add_source("file://resources/patches/openssl/openssl-fips.cnf.patch")
+
+  # Overlay openssl-fips onto puppet-runtime in /opt
+  #
+  # fipsmodule.cnf needs to be generated at install time on the host, so never
+  # include it in the package.
+  #
+  # openssl.cnf also needs to be updated at install time, so don't include it in
+  # the package. Instead ship `openssl-fips.cnf` and overwrite `openssl.cnf`
+  # with that version after the `fips.so` library is installed.
+  #
+  pkg.install do
+    [
+      "gunzip -c #{tarball_name} | #{platform.tar} --skip-old-files -C / -xf -",
+      "rm -f /opt/puppetlabs/puppet/ssl/fipsmodule.cnf",
+      "mv /opt/puppetlabs/puppet/ssl/openssl.cnf /opt/puppetlabs/puppet/ssl/openssl-fips.cnf",
+      "#{platform.patch} --strip=1 --fuzz=0 --ignore-whitespace --no-backup-if-mismatch -d /opt/puppetlabs/puppet/ssl/ < openssl-fips.cnf.patch"
+    ]
+  end
+
+  # See "Making all applications use the FIPS module by default" in
+  # https://www.openssl.org/docs/man3.0/man7/fips_module.html
+  #
+  # We have to run `openssl fipsinstall` to generate `fipsmodule.cnf`. This
+  # generates an HMAC for the `fips.so` and stores that value in `fipsmodule.cnf`.
+  # Only later can we enable fips, by overwriting `openssl.cnf` with the fips
+  # enabled version.
+  #
+  # During an upgrade, `fips.so` may be modified, invaliding the checksum in
+  # `fipsmodule.cnf`, and that will prevent `openssl fipsinstall` from working.
+  # So override OPENSSL_CONF to use the pristine `openssl.cnf` without fips
+  # enabled to regenerate `fipsmodule.conf`
+  #
+  # The `|| :` prevents the install/upgrade from exiting if the command fails.
+  #
+  pkg.add_postinstall_action ["install", "upgrade"],
```

Comment:
I would probably add preremove_action to clean up the generated files: fipsmodule.cnf && openssl.cnf

During the postinstall, I would also suggest that the same have their ownership and file permissions carefully set to the minimum needed. `openssl fipsinstall` may do this but I would still explicitly do it here as well.

Reply:
I added a `preremove_action` and explicitly `chmod` both files to `0644`, which matches what I see on other distributions, e.g. 

```
$ ls -la /etc/ssl/openssl.cnf           
-rw-r--r-- 1 root 11K Apr 20  2020 /etc/ssl/openssl.cnf
```